File: src/lib/ai/providers/gemini.ts
// Copyright (C) 2026 AnalyseDeCircuit
// SPDX-License-Identifier: GPL-3.0-only

/**
 * Google Gemini Provider Adapter
 *
 * Supports Google's Generative Language API with SSE streaming.
 * Uses the `generateContent` endpoint with `streamGenerateContent`.
 */

import type { AiStreamProvider, AiRequestConfig, ChatMessage, AiStreamEvent, AiToolDefinition } from '../providers';
import { aiFetch, aiFetchStreaming } from '../aiFetch';

/**
 * Convert standard ChatMessage format to Gemini API format.
 * Handles tool role by converting to functionResponse parts.
 */
function convertMessages(messages: ChatMessage[]): {
  systemInstruction: string | undefined;
  contents: Array<{ role: 'user' | 'model'; parts: Array<Record<string, unknown>> }>;
} {
  let systemInstruction: string | undefined;
  const contents: Array<{ role: 'user' | 'model'; parts: Array<Record<string, unknown>> }> = [];

  for (const msg of messages) {
    if (msg.role === 'system') {
      systemInstruction = systemInstruction
        ? `${systemInstruction}\n\n${msg.content}`
        : msg.content;
    } else if (msg.role === 'tool') {
      // Gemini uses functionResponse in a user-role message
      let response: Record<string, unknown> = { output: msg.content };
      try { response = JSON.parse(msg.content); } catch { /* use as string */ }
      contents.push({
        role: 'user',
        parts: [{ functionResponse: { name: msg.tool_name || 'unknown', response } }],
      });
    } else if (msg.role === 'assistant' && msg.tool_calls && msg.tool_calls.length > 0) {
      // Model message with function calls
      const parts: Array<Record<string, unknown>> = [];
      if (msg.content) parts.push({ text: msg.content });
      for (const tc of msg.tool_calls) {
        let args: Record<string, unknown> = {};
        try { args = JSON.parse(tc.arguments); } catch { /* empty */ }
        parts.push({ functionCall: { name: tc.name, args } });
      }
      contents.push({ role: 'model', parts });
    } else {
      const role = msg.role === 'assistant' ? 'model' : 'user';
      const last = contents[contents.length - 1];
      // Gemini requires alternating roles
      if (last && last.role === role) {
        last.parts.push({ text: msg.content });
      } else {
        contents.push({ role, parts: [{ text: msg.content }] });
      }
    }
  }

  // Ensure starts with user
  if (contents.length > 0 && contents[0].role !== 'user') {
    contents.unshift({ role: 'user', parts: [{ text: '(Continue)' }] });
  }

  return { systemInstruction, contents };
}

/**
 * Convert AiToolDefinition[] to Gemini functionDeclarations format.
 */
function convertTools(tools: AiToolDefinition[]): Array<{ functionDeclarations: Array<{ name: string; description: string; parameters: Record<string, unknown> }> }> {
  return [{
    functionDeclarations: tools.map((t) => ({
      name: t.name,
      description: t.description,
      parameters: t.parameters,
    })),
  }];
}

function applyToolChoice(body: Record<string, unknown>, config: AiRequestConfig): void {
  if (!config.tools || config.tools.length === 0 || !config.toolChoice || config.toolChoice === 'auto') {
    return;
  }

  const functionCallingConfig: Record<string, unknown> = { mode: 'ANY' };
  if (config.toolChoice !== 'required') {
    functionCallingConfig.allowedFunctionNames = [config.toolChoice.name];
  }
  body.toolConfig = { functionCallingConfig };
}

export const geminiProvider: AiStreamProvider = {
  type: 'gemini',
  displayName: 'Google Gemini',

  async *streamCompletion(
    config: AiRequestConfig,
    messages: ChatMessage[],
    signal: AbortSignal
  ): AsyncGenerator<AiStreamEvent> {
    const cleanBaseUrl = config.baseUrl.replace(/\/+$/, '');
    // Gemini uses API key as query param, not Bearer token
    const url = `${cleanBaseUrl}/models/${encodeURIComponent(config.model)}:streamGenerateContent?alt=sse&key=${encodeURIComponent(config.apiKey)}`;

    const { systemInstruction, contents } = convertMessages(messages);

    const body: Record<string, unknown> = { contents };
    if (systemInstruction) {
      body.system_instruction = { parts: [{ text: systemInstruction }] };
    }
    if (config.maxResponseTokens) {
      body.generationConfig = { maxOutputTokens: config.maxResponseTokens };
    }
    if (config.tools && config.tools.length > 0) {
      body.tools = convertTools(config.tools);
      applyToolChoice(body, config);
    }

    const { response: statusPromise, body: streamBody } = aiFetchStreaming(url, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify(body),
      signal,
    });

    const { ok, status } = await statusPromise;

    if (!ok) {
      const errReader = streamBody.getReader();
      const errDecoder = new TextDecoder();
      let errorText = '';
      try {
        while (true) {
          const { done, value } = await errReader.read();
          if (done) break;
          errorText += errDecoder.decode(value, { stream: true });
        }
      } catch { /* stream error */ }
      let errorMessage = `Gemini API error: ${status}`;
      try {
        const errorJson = JSON.parse(errorText);
        errorMessage = errorJson.error?.message || errorMessage;
      } catch {
        if (errorText) errorMessage = errorText.slice(0, 200);
      }
      yield { type: 'error', message: errorMessage };
      return;
    }

    const reader = streamBody.getReader();
    if (!reader) {
      yield { type: 'error', message: 'No response body' };
      return;
    }

    const decoder = new TextDecoder();
    let buffer = '';

    const processDataLine = (line: string): AiStreamEvent[] => {
      if (!line.startsWith('data: ')) return [];
      const data = line.slice(6).trim();
      if (!data) return [];

      const events: AiStreamEvent[] = [];

      try {
        const json = JSON.parse(data);
        const candidates = json.candidates;
        if (candidates?.[0]?.content?.parts) {
          for (const part of candidates[0].content.parts) {
            if (part.text) {
              events.push({ type: 'content', content: part.text });
            }
            if (part.functionCall) {
              const callId = `gemini-${Date.now()}-${Math.random().toString(36).slice(2, 7)}`;
              const args = JSON.stringify(part.functionCall.args || {});
              events.push({ type: 'tool_call_complete', id: callId, name: part.functionCall.name, arguments: args });
            }
          }
        }
      } catch {
        // Ignore parse errors
      }

      return events;
    };

    try {
      while (true) {
        if (signal.aborted) break;
        const { done, value } = await reader.read();
        if (done) break;

        buffer += decoder.decode(value, { stream: true });
        const lines = buffer.split('\n');
        buffer = lines.pop() || '';

        for (const line of lines) {
          for (const event of processDataLine(line)) {
            yield event;
          }
        }
      }

      if (buffer.trim()) {
        for (const event of processDataLine(buffer.trim())) {
          yield event;
        }
      }
    } finally {
      reader.releaseLock();
    }

    yield { type: 'done' };
  },

  async fetchModels(config: { baseUrl: string; apiKey: string }): Promise<string[]> {
    const cleanBaseUrl = config.baseUrl.replace(/\/+$/, '');
    const resp = await aiFetch(
      `${cleanBaseUrl}/v1beta/models?key=${config.apiKey}`
    );
    if (!resp.ok) throw new Error(`Failed to fetch models: ${resp.status}`);
    const data = JSON.parse(resp.body);
    if (!Array.isArray(data.models)) return [];
    return data.models
      .filter((m: { supportedGenerationMethods?: string[] }) =>
        m.supportedGenerationMethods?.includes('generateContent')
      )
      .map((m: { name: string }) => m.name.replace('models/', ''))
      .sort();
  },

  async fetchModelDetails(config: { baseUrl: string; apiKey: string }): Promise<Record<string, number>> {
    const cleanBaseUrl = config.baseUrl.replace(/\/+$/, '');
    const resp = await aiFetch(
      `${cleanBaseUrl}/v1beta/models?key=${config.apiKey}`
    );
    if (!resp.ok) return {};
    const data = JSON.parse(resp.body);
    if (!Array.isArray(data.models)) return {};
    const result: Record<string, number> = {};
    for (const m of data.models) {
      // Gemini returns inputTokenLimit
      const ctx = m.inputTokenLimit;
      const id = m.name?.replace('models/', '') || '';
      if (typeof ctx === 'number' && ctx > 0 && id) {
        result[id] = ctx;
      }
    }
    return result;
  },
};

File: crates/oxideterm-ai/src/streaming/gemini.rs
use anyhow::{Context, Result, anyhow};
use serde_json::Value;
use std::collections::HashMap;
use std::sync::atomic::{AtomicU64, Ordering};

use crate::providers::{api_key_required_ref, url_encode_component};
use crate::{
    AiChatMessage, AiChatRole, AiChatStreamConfig, AiStreamEvent, AiToolCall, AiToolChoice,
    AiToolDefinition,
};

use super::CHAT_STREAM_TIMEOUT;
use super::common::{ParsedStreamLine, stream_sse_response};

static GEMINI_TOOL_CALL_COUNTER: AtomicU64 = AtomicU64::new(1);

pub(crate) async fn stream_gemini_completion(
    config: AiChatStreamConfig,
    messages: Vec<AiChatMessage>,
    events: tokio::sync::mpsc::UnboundedSender<AiStreamEvent>,
) -> Result<()> {
    let api_key = api_key_required_ref(&config.provider_type, config.api_key.as_ref())?;
    let url = format!(
        "{}/models/{}:streamGenerateContent",
        config.base_url.trim().trim_end_matches('/'),
        url_encode_component(&config.model)
    );
    let client = reqwest::Client::builder()
        .timeout(CHAT_STREAM_TIMEOUT)
        .build()
        .context("failed to create Gemini chat client")?;
    let body = gemini_chat_body(&config, &messages);
    let response = client
        .post(&url)
        // Gemini requires the API key as a query parameter. Let reqwest attach
        // it to the request and strip URLs from transport errors below.
        .query(&[("alt", "sse"), ("key", api_key.as_str())])
        .header(reqwest::header::CONTENT_TYPE, "application/json")
        .json(&body)
        .send()
        .await
        .map_err(|error| {
            anyhow!(
                "failed to connect to Gemini provider: {}",
                error.without_url()
            )
        })?;
    if !response.status().is_success() {
        let status = response.status().as_u16();
        let error_text = response.text().await.unwrap_or_default();
        return Err(anyhow!(parse_gemini_error(status, &error_text)));
    }
    let _ = stream_sse_response(response, &events, parse_gemini_data_line).await?;
    let _ = events.send(AiStreamEvent::Done);
    Ok(())
}

pub(crate) fn gemini_chat_body(config: &AiChatStreamConfig, messages: &[AiChatMessage]) -> Value {
    let (system_instruction, contents) = gemini_chat_contents(messages);
    let mut body = serde_json::json!({ "contents": contents });
    if let Some(system) = system_instruction
        && let Some(object) = body.as_object_mut()
    {
        object.insert(
            "system_instruction".to_string(),
            serde_json::json!({ "parts": [{ "text": system }] }),
        );
    }
    if let Some(tokens) = config.max_response_tokens.filter(|tokens| *tokens > 0)
        && let Some(object) = body.as_object_mut()
    {
        object.insert(
            "generationConfig".to_string(),
            serde_json::json!({ "maxOutputTokens": tokens }),
        );
    }
    if !config.tools.is_empty()
        && let Some(object) = body.as_object_mut()
    {
        object.insert(
            "tools".to_string(),
            serde_json::json!(gemini_tool_definitions(&config.tools)),
        );
        if let Some(tool_config) = gemini_tool_config(&config.tool_choice) {
            object.insert("toolConfig".to_string(), tool_config);
        }
    }
    body
}

pub(crate) fn gemini_chat_contents(messages: &[AiChatMessage]) -> (Option<String>, Vec<Value>) {
    let mut system_parts = Vec::new();
    let mut contents = Vec::<Value>::new();
    let mut tool_names_by_id = HashMap::<String, String>::new();
    for message in messages {
        match message.role {
            AiChatRole::System if !message.content.is_empty() => {
                system_parts.push(message.content.clone());
            }
            AiChatRole::System => {}
            AiChatRole::Tool => {
                let name = message
                    .tool_call_id
                    .as_ref()
                    .and_then(|id| tool_names_by_id.get(id))
                    .cloned()
                    .unwrap_or_else(|| "unknown".to_string());
                let response = serde_json::from_str::<Value>(&message.content)
                    .unwrap_or_else(|_| serde_json::json!({ "output": message.content }));
                contents.push(serde_json::json!({
                    "role": "user",
                    "parts": [{ "functionResponse": { "name": name, "response": response } }],
                }));
            }
            AiChatRole::Assistant if !message.tool_calls.is_empty() => {
                let mut parts = Vec::new();
                if !message.content.is_empty() {
                    parts.push(serde_json::json!({ "text": message.content }));
                }
                for call in message.tool_calls.iter().filter_map(AiToolCall::from_value) {
                    tool_names_by_id.insert(call.id.clone(), call.name.clone());
                    let args = serde_json::from_str::<Value>(&call.arguments)
                        .ok()
                        .filter(Value::is_object)
                        .unwrap_or_else(|| serde_json::json!({}));
                    parts.push(serde_json::json!({
                        "functionCall": { "name": call.name, "args": args },
                    }));
                }
                contents.push(serde_json::json!({ "role": "model", "parts": parts }));
            }
            AiChatRole::User | AiChatRole::Assistant => {
                let role = if message.role == AiChatRole::Assistant {
                    "model"
                } else {
                    "user"
                };
                if let Some(last) = contents.last_mut()
                    && last.get("role").and_then(Value::as_str) == Some(role)
                    && let Some(parts) = last.get_mut("parts").and_then(Value::as_array_mut)
                {
                    parts.push(serde_json::json!({ "text": message.content }));
                    continue;
                }
                contents.push(serde_json::json!({
                    "role": role,
                    "parts": [{ "text": message.content }],
                }));
            }
        }
    }
    if contents
        .first()
        .is_some_and(|content| content.get("role").and_then(Value::as_str) != Some("user"))
    {
        contents.insert(
            0,
            serde_json::json!({ "role": "user", "parts": [{ "text": "(Continue)" }] }),
        );
    }
    let system = (!system_parts.is_empty()).then(|| system_parts.join("\n\n"));
    (system, contents)
}

fn gemini_tool_definitions(tools: &[AiToolDefinition]) -> Vec<Value> {
    vec![serde_json::json!({
        "functionDeclarations": tools
            .iter()
            .map(|tool| serde_json::json!({
                "name": tool.name,
                "description": tool.description,
                "parameters": tool.parameters,
            }))
            .collect::<Vec<_>>(),
    })]
}

fn gemini_tool_config(tool_choice: &AiToolChoice) -> Option<Value> {
    match tool_choice {
        AiToolChoice::Auto => None,
        AiToolChoice::Required => Some(serde_json::json!({
            "functionCallingConfig": { "mode": "ANY" },
        })),
        AiToolChoice::Named(name) if !name.is_empty() => Some(serde_json::json!({
            "functionCallingConfig": {
                "mode": "ANY",
                "allowedFunctionNames": [name],
            },
        })),
        AiToolChoice::Named(_) => None,
    }
}

pub(crate) fn parse_gemini_data_line(line: &str) -> ParsedStreamLine {
    let Some(data) = line.strip_prefix("data: ") else {
        return ParsedStreamLine {
            events: Vec::new(),
            saw_frame: false,
        };
    };
    let data = data.trim();
    if data.is_empty() {
        return ParsedStreamLine {
            events: Vec::new(),
            saw_frame: true,
        };
    }

    let mut events = Vec::new();
    if let Ok(json) = serde_json::from_str::<Value>(data)
        && let Some(parts) = json
            .get("candidates")
            .and_then(Value::as_array)
            .and_then(|candidates| candidates.first())
            .and_then(|candidate| candidate.get("content"))
            .and_then(|content| content.get("parts"))
            .and_then(Value::as_array)
    {
        for part in parts {
            if let Some(text) = part
                .get("text")
                .and_then(Value::as_str)
                .filter(|text| !text.is_empty())
            {
                events.push(AiStreamEvent::Content(text.to_string()));
            }
            if let Some(function_call) = part.get("functionCall") {
                let id = format!(
                    "gemini-{}",
                    GEMINI_TOOL_CALL_COUNTER.fetch_add(1, Ordering::Relaxed)
                );
                let name = function_call
                    .get("name")
                    .and_then(Value::as_str)
                    .unwrap_or("unknown")
                    .to_string();
                let arguments = function_call
                    .get("args")
                    .filter(|args| args.is_object())
                    .cloned()
                    .unwrap_or_else(|| serde_json::json!({}))
                    .to_string();
                events.push(AiStreamEvent::ToolCallComplete {
                    id,
                    name,
                    arguments,
                });
            }
        }
    }
    ParsedStreamLine {
        events,
        saw_frame: true,
    }
}

fn parse_gemini_error(status: u16, body: &str) -> String {
    let mut fallback = format!("Gemini API error: {status}");
    if let Ok(json) = serde_json::from_str::<Value>(body) {
        if let Some(message) = json
            .get("error")
            .and_then(|error| error.get("message"))
            .and_then(Value::as_str)
        {
            return message.to_string();
        }
    } else if !body.is_empty() {
        fallback = body.chars().take(200).collect();
    }
    fallback
}
